switch to pipesys for build input and output

bcressey opened the pull request:
**Issue number:**
N/A

**Description of changes:**
About a year ago, as a weekend hack, I set out to solve the problem of passing make jobserver file descriptors into a build in order to remove the `BUILDSYS_JOBS = "8"` constraint in favor of more dynamic parallelism that would use all the CPUs available on the build host in an intelligent way. `pipesys` was the solution I hit upon, but the performance benefits weren't clear cut, and there were a lot of annoying edge cases across the different `make` invocations in various packages.

While benchmarking, I observed that the greatest impediment to additional parallelism was not efficient use of all available CPU cores, but rather that the Docker daemon struggled under the increased load of additional parallel jobs, with each one copying a slightly different version of a multi-gigabyte build context as input, and compressing the intermediate stages on the way to delivering the package and image outputs.

I realized that if I could pass jobserver file descriptors from `make`, I could also pass directory file descriptors for the project workspace. That provided far more obvious wins even without adding more simultaneous package builds: it sped up builds and reduced `dockerd` CPU and memory usage. I also realized I could pass a file descriptor for the output directory, and avoid the need to spin up a container just to copy out the build artifacts. Accordingly, I redesigned `pipesys` around the use case of build inputs and outputs.

The commits here are best read in reverse order. My ultimate goal was to change the `.dockerignore` file to exclude almost everything from the build context. The Dockerfile changes, while extensive, mostly amount to changing the source of input files from "`/host`" to "`/bypass`", and to write output files directly to "`/output`" rather than leaving that up to Docker. (The most complicated parts of `pipesys` are in the "link" command, and that's done to support the very convenient `Dockerfile` interface.)

The `buildsys` changes are relatively straightforward: we need to serve two file descriptors - one for input, one for output - via `pipesys`. We can also dispense with caching for intermediate layers and switch back to a single `Dockerfile`.

`unplug` is a simple program that offers just the isolation semantics we need - no network access - without  requiring any additional privileges in the build environment.

`pipesys` is complex new code, but may be easier to follow once the specific uses are understood. 

**Testing done:**
Built https://github.com/bottlerocket-os/bottlerocket-core-kit. Before, the build was so flaky that it was difficult to finish building all packages, and the core kit took several minutes to create. After, the build was reliable across a nightly loop of `make twoliter clean && make`, and the core kit takes around 20 seconds to create.


**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

Commit:
pipesys: add tool to share file descriptors

For Bottlerocket, package and image builds both involve copying large
binary files into and out of builds. When using Docker to create the
build environment, all this data must pass through the Docker daemon,
which balloons its resident memory usage in unpredictable ways and
can lead to crashes or OOM kills.

Docker also implements caching for intermediate layers. While useful
in the general case, this is unhelpful for these builds, because a
minor edit for troubleshooting purposes can produce gigabytes of data
that will never be useful again but that remain in Docker's cache.

Docker understandably does not provide a straightforward way to bind-
mount a host directory into a build, as that would decouple the build
from local inputs that are potentially required for it to succeed,
making it difficult for others to reproduce the build. Since buildsys
manages those inputs for Bottlerocket builds, this is not as much of
a concern, but the limitation remains.

On Linux, Unix domain sockets can be used to pass file descriptors.
They can also be abstract, with no backing socket file, and only
accessible within the network namespace of the opening process. This
can be used to bend Docker's rules and provide the equivalent of a
bind mount, with enough effort.

pipesys provides an abstraction over these capabilities, in the form
of library code to serve and fetch file descriptors. It can run as a
standalone program to serve a path's file descriptor to any clients
that send a request. It can also work as a bind mount substitute, by
retrieving a file descriptor and creating a symlink to it for easy
access from subsequent programs.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
unplug: add tool to restrict network access

For pipesys to provide file descriptor accesses via an abstract UNIX
domain socket into container builds, the build must run with in the
host network namespace.

That conflicts with the requirement that packages are built with no
access to the network. This ensures that builds cannot have unmodeled
dependencies on remote resources that could cause them to break or
produce different results across executions.

To satisfy the requirement, create an "unplug" tool that blocks most
socket-related calls. Existing solutions like firejail or nsjail can
also do this, but they expect to run in a privileged context in order
to set up namespaces and the like, and build containers do not have
these privileges.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
buildsys: share input and output dirs via pipesys

Rather than copying source archives and other large files into the
build container via the build context, use pipesys to serve a read-
only copy of the project root directory so that the build can access
any files it needs directly.

For output, provide a file descriptor for the output directory via
pipesys rather than starting a container from the image and copying
the files out.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
twoliter: include pipesys and unplug tools

pipesys is used to pass file descriptors for input and output dirs,
and unplug is used to block network access during package builds.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
twoliter: use pipesys for I/O in build Dockerfile

buildsys provides the bypass socket to copy input files from a read-
only bind mount of the project's workspace, and the output socket to
share the expected output directory.

Use the file descriptors served by these sockets for I/O to and from
the build host across all package, kit, and image builds.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
twoliter: use pipesys for I/O in repack Dockerfile

buildsys provides the bypass socket to copy input files from a read-
only bind mount of the project's workspace, and the output socket to
share the expected output directory.

Use the file descriptors served by these sockets for I/O to and from
the build host during image repack operations.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
twoliter: combine build and repack Dockerfiles

The build and repack Dockerfile were separate because repack needs
access to the disk and filesystem image files that were too costly
to copy through the build context for all builds.

Now that bulk binary data is passed through pipesys, the stages for
repack can be merged back into the primary Dockerfile.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
buildsys: use build.Dockerfile for all builds

All build targets have moved to build.Dockerfile.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
buildsys: do not cache intermediate build stages

The intermediate build stages are not helpful - if cargo has decided
that something should be rebuilt, then reusing a cached layer from a
previous build could invalidate the result.

The Dockerfile echoes the NOCACHE build argument periodically to bust
any caching that might otherwise happen. However, there are certain
cases where caching is helpful, such as the cache layer provided to
Rust builds.

BuildKit supports the "no-cache-filter" option to allow caching to be
selectively disabled for specific intermediate stages. Add all the
intermediate stages in build.Dockerfile to that list.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

bcressey commented:
Definitely curious to see if this works as expected in the CI environment!

bcressey commented:
The irony is not lost on me that `unplug` - what I called the "simple program" - was responsible for the above two force pushes. Just to add the license statement and appease `cargo fmt`, but still not a promising start.

bcressey commented:
⬆️ force push fixes a couple of bugs with variant builds.

The more significant change is the addition of `rm -rf /output/*` to the various stages that interact with the output directory. What's going on here is that if a build fails, it's possible for the directory to contain root-owned artifacts that were partly written and not yet `chown`'d by one of the parallel scripts (`rpm2img`, `rpm2migration`, `rpm2kmodkit`).

These artifacts are slightly annoying for `buildsys` to clean up because it is (presumably) running as a non-root user, which defeats a simple `std::fs::remove_dir_all`.

Doing it in the Dockerfile means that clean up can happen as root, and then all of the changes as far as adding, removing, and chmod'ing contents happen in the same place.

bcressey commented:
Oh but now there is a bug in the `imgbuild` stage since the three `rm -rf /output/*` calls are unsynchronized; if one stage runs significantly later than the others, it may remove content that's been newly written.

Fix incoming ...

cbgbt commented on tools/buildsys/src/builder.rs:
```diff
@@ -571,31 +578,71 @@ impl DockerBuild {
-        let cp = format!("cp {}:/output/. {}", self.tag, marker_dir.display()).split_string();
-        let rm = format!("rm --force {}", self.tag).split_string();
-        let rmi = format!("rmi --force {}", self.tag).split_string();
+        // Run a container with the project's root as a read-only volume mount, so that pipesys can
+        // serve a read-only file descriptor that's safe to pass into builds.
```
Cool idea.

cbgbt commented on tools/pipesys/src/cmd/mod.rs:
```diff
@@ -0,0 +1,103 @@
+}
+
+/// Helper function to retrieve file descriptors via an abstract socket.
+fn fetch_fds(socket: &str, wanted: usize) -> Result<Vec<i32>> {
+    let addr = uds::UnixSocketAddr::from_abstract(socket.as_bytes())
+        .with_context(|| format!("failed to create socket {}", socket))?;
```
Seems like `wanted` doesn't actually behave like expected. Should probably alter the function or else just always fetch one FD as a constant.

cbgbt commented on tools/pipesys/src/cmd/link.rs:
```diff
@@ -0,0 +1,194 @@
+    /// Retrieve the file descriptor, then spawn a background process to create the link so that it
+    /// survives the return of the foreground process.
+    pub(crate) async fn execute(&self) -> Result<()> {
+        // Remove the existing symlink, if present.
+        let _ = fs::remove_file(&self.target).await;
```
Should we be more defensive when we remove the target?

I'm not sure if we can assert that it is a pipesys-managed link here, but we could maybe refuse to delete non-symlinks.

cbgbt commented on tools/pipesys/src/cmd/mod.rs:
```diff
@@ -0,0 +1,103 @@
+
+    let mut fd_buf = [-1; 8];
+    let (_, _, fds) = client
+        .recv_fds(&mut [0u8; 1], &mut fd_buf)
```
I didn't know that Linux could do this. That's awesome.

cbgbt commented on tools/pipesys/src/cmd/link.rs:
```diff
@@ -0,0 +1,194 @@
+where
+    F: Future<Output = bool> + 'a,
+{
+    // Check whether the provided function succeeds right away, in case the watched file was
+    // created or removed before the inotify watch could be started.
+    if check(path).await {
+        return Ok(());
+    }
```
Do we need to perform this check after the `event_stream` is created? I'm imagining a very short race window otherwise.

bcressey replied:
> Do we need to perform this check after the event_stream is created? I'm imagining a very short race window otherwise.

My understanding is that the kernel will publish events to the watch descriptor as soon as the `inotify.watches.add(...)` returns. On the Rust side, we won't start reading those events until we poll the event stream, but we shouldn't miss any.

cbgbt replied:
When I reviewed that function I forgot that it was called in the context of already having called `inotify_init`. Makes sense!

cbgbt reviewed (commented):
Nice!

The multi-fd support in pipesys would be more useful if there was some scheme for providing abstract names to the descriptors so that clients could understand what the returned descriptors mapped to.

Since we're only interested in serving a single descriptor per-process, maybe we should reduce the scope of the program to focus on one descriptor?

bcressey commented:
> Since we're only interested in serving a single descriptor per-process, maybe we should reduce the scope of the program to focus on one descriptor?

Sounds good. The multi-fd stuff is left over from the initial goal of passing `make`'s jobserver fds, where there are two (the read & write ends of a pipe).

webern commented on tools/pipesys/src/server.rs:
```diff
@@ -0,0 +1,114 @@
+use uds::{tokio::UnixSeqpacketListener, UnixSocketAddr};
+
+#[derive(Clone, Debug, Parser)]
+pub struct Server {
```
```suggestion

/// Describe what the program does.
#[derive(Clone, Debug, Parser)]
pub struct Server {
```

webern commented on tools/buildsys/src/builder.rs:
```diff
@@ -25,6 +26,7 @@ use std::collections::HashSet;
 use std::env;
 use std::fs::{self, read_dir, File};
 use std::num::NonZeroU16;
+use std::os::unix::fs::MetadataExt;
```
Things start becoming interesting (I hadn't thought of this before), if `buildsys` needs to be compiled for Linux (because it runs in a container, e.g. if cargo make and cargo are used from a "bootstrap" container, as I used to call it.) and `twoliter` runs on a host that isn't Linux.

bcressey replied:
> Things start becoming interesting (I hadn't thought of this before), if buildsys needs to be compiled for Linux ...

Good point. It doesn't seem likely that cargo's bindeps will ever support building a binary for a different target than the crate is being built for, so we'd need to approach the build rather differently to get a Mac or Windows binary with Linux tools embedded.

bcressey replied:
For now I've gone with the approach of making everything build on non-Linux, while leaving the actual logic unimplemented for `pipesys` and `unplug`. That way it's still possible to develop Twoliter on a different OS, though building kits or variants will be Linux-only.

webern commented on twoliter/embedded/build.Dockerfile.dockerignore:
```diff
@@ -1,17 +1,16 @@
+
+# Include the specific files expected by the Dockerfile.
+!/build/tools
+!/packages/*/*.spec
```
Why are these unignored but, for example, tar.gz files are not?

bcressey replied:
> Why are these unignored but, for example, tar.gz files are not?

Spec files are tiny, and - like files under sources - are somewhat likely to be edited while a build is in progress.

Including these in the build context has the effect of snapshotting them at the moment when the build starts (more or less); copying them from the bypass directory would mean that edits would go "live" in a way that could lead to confusing build errors.

webern commented on tools/pipesys/src/cmd/link.rs:
```diff
@@ -0,0 +1,194 @@
+        if let Some(event_file_name) = event.name {
+            if event_file_name == file_name {
+                trace!("found event for {}", path.display());
+                if check(path).await {
```
Is this `fs`?

```suggestion
                if fs::check(path).await {
```

bcressey replied:
> Is this `fs`?

`check` is a function parameter - maybe it needs a better name?

webern commented on tools/unplug/src/main.rs:
```diff
@@ -0,0 +1,60 @@
+        return Ok(code.into());
+    }
+
+    Ok(0.into())
```
nit, since most of what you're doing is way beyond me.

```suggestion
    Ok(ExitCode::Success)
```

webern commented on twoliter/embedded/build.Dockerfile:
```diff
@@ -65,46 +65,53 @@ RUN \
         ln -snft ./rpmbuild/RPMS {} \+ ; \
     done && \
     createrepo_c \
-        -o ./rpmbuild/RPMS \
```
nit: but making the diff bigger than it needs to be is probably not the best in a complicated PR like this. I would suggest re-format the Dockerfile in a subsequent PR.

webern commented on twoliter/embedded/repack.Dockerfile:
```diff
@@ -1,69 +0,0 @@
-# syntax=docker/dockerfile:1.4.3
```
This commit seems to be missing the delete of repack.Dockerfile.dockerignore `twoliter: combine build and repack Dockerfiles`

webern commented on twoliter/embedded/repack.Dockerfile.dockerignore:
```diff
@@ -1,4 +0,0 @@
-/
```
Maybe it's just a git thing, but this file delete is in the commit after the one that removed it from Twoliter's embedding.

Commit:
twoliter: refactor build.Dockerfile.dockerignore

When using pipesys for inputs and outputs, most of the build context
can be excluded. The exceptions are source files that are useful to
snapshot at the moment the build is initiated, such as package specs
and first-party code under sources, along with tools such as pipesys
itself that must be present in the build.

Current versions of `cargo` do not work with a completely read-only
copy of the crates in `CARGO_HOME`, so that directory must be sent
via build context even though it is nominally read-only.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

cbgbt reviewed (approved):
Looks good, though I'm not sure I see how you resolved this:
> Oh but now there is a bug in the imgbuild stage since the three rm -rf /output/* calls are unsynchronized; if one stage runs significantly later than the others, it may remove content that's been newly written.

Does it have something to do with the modified stage dependencies?

bcressey commented:
> Does it have something to do with the modified stage dependencies?

Now the output cleanup happens just once in the `repobuild` stage, which is used by the other three stages for building a variant (`imgbuild`, `kmodkitbuild`, `migrationbuild`). `migrationbuild` now depends on `repobuild` like the other two.

sam-berning commented on tools/pipesys/src/cmd/mod.rs:
```diff
@@ -0,0 +1,117 @@
+#[cfg(target_os = "linux")]
+use anyhow::{ensure, Context};
+use clap::Parser;
+use env_logger::Builder;
```
nit: I think it'd be slightly more readable not to `use` this since it's such a generic name

bcressey commented:
⬆️ force push switches to `env_logger::Builder` rather than importing it.
